firewood/src/merkle/tests/proof.rs

```diff
@@ -0,0 +1,271 @@
+}
+
+#[test]
+fn test_proof() {
+    let rng = firewood_storage::SeededRng::from_env_or_random();
+    let set = fixed_and_pseudorandom_data(&rng, 500);
+    let mut items = set.iter().collect::<Vec<_>>();
+    items.sort_unstable();
+    let merkle = init_merkle(items.clone());
+
+    let root_hash = merkle.nodestore().root_hash().unwrap();
+
+    for (key, val) in items {
+        let proof = merkle.prove(key).unwrap();
+        assert!(!proof.is_empty());
+        proof.verify(key, Some(val), &root_hash).unwrap();
+    }
+}
+
+#[test]
+fn test_proof_end_with_leaf() {
+    let merkle = init_merkle([
+        ("do", "verb"),
+        ("doe", "reindeer"),
+        ("dog", "puppy"),
+        ("doge", "coin"),
+        ("horse", "stallion"),
+        ("ddd", "ok"),
+    ]);
+    let root_hash = merkle.nodestore().root_hash().unwrap();
+
+    let key = b"doe";
+
+    let proof = merkle.prove(key).unwrap();
+    assert!(!proof.is_empty());
+
+    proof.verify(key, Some(b"reindeer"), &root_hash).unwrap();
+}
+
+#[test]
+fn test_proof_end_with_branch() {
+    let items = [
+        ("d", "verb"),
+        ("do", "verb"),
+        ("doe", "reindeer"),
+        ("e", "coin"),
+    ];
+    let merkle = init_merkle(items);
+    let root_hash = merkle.nodestore().root_hash().unwrap();
+
+    let key = b"d";
+
+    let proof = merkle.prove(key).unwrap();
+    assert!(!proof.is_empty());
+
+    proof.verify(key, Some(b"verb"), &root_hash).unwrap();
+}
+
+#[test]
+fn test_bad_proof() {
+    let rng = firewood_storage::SeededRng::from_env_or_random();
+    let set = fixed_and_pseudorandom_data(&rng, 800);
+    let mut items = set.iter().collect::<Vec<_>>();
+    items.sort_unstable();
+    let merkle = init_merkle(items.clone());
+    let root_hash = merkle.nodestore().root_hash().unwrap();
+
+    for (key, value) in items {
+        let proof = merkle.prove(key).unwrap();
+        assert!(!proof.is_empty());
+
+        // Delete an entry from the generated proofs.
+        let mut new_proof = proof.into_mutable();
+        new_proof.pop();
+
+        // TODO: verify error result matches expected error
+        assert!(new_proof.verify(key, Some(value), &root_hash).is_err());
+    }
+}
+
+#[test]
+fn exclusion_with_proof_value_present() {
+    // Build a trie where an ancestor on the path has a value
+    let mut merkle = crate::merkle::tests::create_in_memory_merkle();
+    // Parent has a value
+    merkle.insert(&[0u8], Box::from([0u8])).unwrap();
+    // Child under the same branch
+    merkle.insert(&[0u8, 1u8], Box::from([1u8])).unwrap();
+
+    let merkle = merkle.hash();
+    let root_hash = merkle.nodestore.root_hash().unwrap();
+    debug!("{}", merkle.dump_to_string().unwrap());
+    debug!("root_hash: {root_hash:?}");
+
+    // Non-existent key under the same parent branch
+    let missing = [0u8, 2u8];
+    let proof = merkle.prove(&missing).unwrap();
+
+    debug!("{proof:#?}");
+    assert_eq!(
+        proof.as_ref().first().unwrap().to_hash(),
```

Comment:
test fails here with ethhash cause this is too small to be hashed into the triehash:
```suggestion
        TrieHash::from(proof.as_ref().first().unwrap().to_hash()),
```